Feat: port to stm32f0xx

h4kyu opened the pull request:
### Summary
The PR ports the CMSIS USB CDC and HID implementation to the stm32f0xx MCU family. The code is configured specifically for stm32f042x6 chips, but can easily be reconfigured for any of the stm32f0xx chips. 
### Changes
1. Replaced MCU header files:
- `stm32g441xx.h` --> `stm32f042x6.h`
- `stm32g4xx.h` --> `stm32f0xx.h`
- `system_stm32g4xx.h/c` --> `system_stm32f0xx.h/c`

MCU header files for the stm32f0xx family can be found [here](https://github.com/STMicroelectronics/STM32CubeF0).

2. Modified clock initialization to use internal `HSI48` with `CRS` for USB:
- See `InitClock()` in `main.c`
- removed usage of `G4`-specific RCC registers

3. Replaced OTG-style USB interrupts by merging USB ISRs:
- merged `USB_LP_IRQn` and `USB_HP_IRQn` to `USB_IRQn`
- implemented `USB_IRQHandler()`, which calls existing `USB_LP_Handler()`

stm32f042 has the USB Device FS peripheral, not OTS FS. It exposes a single interrupt. 

4. Removed NCM USB class:
- deleted all LWIP dependencies

LWIP requires **8-12 KB SRAM**, above the 6 KB SRAM of stm32f042 devices. 

### Reconfigure for other stm32f0xx

Replace the `stm32f042x6.h` MCU header with the appropriate header for your stm32f0xx chip. If using the STM32CubeIDE, ensure the appropriate symbol for your specific MCU is defined in the project (ex. Symbol `STM32F042x6`, Value `STM32F042x6`). (Properties > C/C++ General > Paths and Symbols > Symbols > Add...)

CShark commented:
Thanks for your contribution & supporting the F042.

The code should still be compatible with the G4 variant though. That means:
- The G4 headers in CMSIS need to stay
- The Include part in main.h is better written as
```
#define <STM Variant> (that part could even be in a separate header file, because that info is needed in multiple files)

#ifdef STM32G441xx
#include <G4-Header>
#elif STM32F042x6
#include <F0-Header>
#endif
```
This way it is easier to switch between the G4 and F0 variant

- The NCM needs to stay. This is a reference repository. Even if someone on an F0 can't use NCM, it is still a reference implementation. If it is not called from the main.c, it won't be included in the memory footprint. Same for LWIP — as long as you don't call them, they won't contribute to the code size
- That also means adding the compile exlusions for LWIPs folders back (test, apps/mqtt, apps/smtp, apps/tftp, apps/http, apps/sntp)
- You can just comment out the initialisation of NCM in main.c or put it into a `#ifndef <G4>` block for your F0. For reference purposes, the code should stay
- The InitClock needs to work with the G4 and the F0. I'd propose two functions (`InitClock_G4`, `InitClock_F0`) which are called from the InitClock using `#ifdef <STMVariant>`
- The `system_stm32g4xx.c` needs to be kept, as well as the startup & linker script for the G4. You can add an `#ifdef` for the correct platform if it doesn't compile correctly.
- The `eth` folder needs to be kept. As long as it isn't called, it won't contribute to the footprint.

Once you can compile the changes for the F0 as well as the G4 and NCM is kept I'll happily merge the changes.

h4kyu commented:
Yup, that makes sense. I'll add NCM and G4 headers back under the preprocessor directives you mentioned, thanks for the detailed response!

h4kyu committed:
preprocessor directives for conditional compilation of clock initialization, mcu header files

h4kyu commented:
I've added the preprocessor directives you mentioned, let me know if here are further issues.

CShark committed:
Changed build system to cmake

- Changed the build system to cmake to have proper support for multiple chip targets
- preferred editor is now vscode, STM32CubeIDE is no longer supported
